lib/ukfallocbuddy/tests/test_ukfallocbuddy.c

```diff
@@ -0,0 +1,217 @@
+#include <uk/test.h>
+#include <uk/fallocbuddy.h>
+#include <stdint.h>
+#include <stdio.h>
+#include <stdlib.h>
+
+struct test_ukfallocbuddy {
+
+	struct uk_falloc *fa;
+	__paddr_t start;
+	__sz len;
+	unsigned long pages;
+	__sz fa_meta_size;
+	void *start_zone;
+	void *fa_meta;
+};
+
+
+static struct uk_falloc *init_new_fb_alloc(void *fa_meta, void *start_zone,
+	unsigned long len, __paddr_t *new_start, __sz fa_meta_size,
+	unsigned long *pages)
+{
+	__vaddr_t dm_off;
+	__paddr_t start = (__paddr_t)start_zone;
+	struct uk_falloc *fa;
+
+	UK_ASSERT(fa = malloc(uk_fallocbuddy_size()));
+
+	UK_ASSERT(!uk_fallocbuddy_init(fa));
+	start = PAGE_ALIGN_UP(start + fa_meta_size);
+	*pages = (len - fa_meta_size) >> PAGE_SHIFT;
+	dm_off = start;
+	UK_ASSERT(!fa->addmem(fa, fa_meta, start, *pages, dm_off));
+	*new_start = start;
+	return fa;
+}
+
+static struct test_ukfallocbuddy *initialize_parameters(__sz len)
+{
+	struct test_ukfallocbuddy *test_attributes =
+(struct test_ukfallocbuddy *) malloc(sizeof(struct test_ukfallocbuddy));
+
+	UK_ASSERT(test_attributes);
+
+	test_attributes->len = len;
+	test_attributes->pages = len >> PAGE_SHIFT;
+	test_attributes->fa_meta_size =
+		uk_fallocbuddy_metadata_size(test_attributes->pages);
+	test_attributes->start_zone = malloc(len);
+	UK_ASSERT(test_attributes->start_zone);
+	test_attributes->fa_meta = malloc(test_attributes->fa_meta_size);
+	UK_ASSERT(test_attributes->fa_meta_size);
+	test_attributes->fa = init_new_fb_alloc(test_attributes->fa_meta,
+						test_attributes->start_zone,
+						test_attributes->len,
+						&(test_attributes->start),
+						test_attributes->fa_meta_size,
+						&(test_attributes->pages));
+	UK_ASSERT(test_attributes->fa);
+
+	return test_attributes;
+
+}
+
+static void free_memory(struct test_ukfallocbuddy *test_attr)
+{
+	free(test_attr->start_zone);
+	free(test_attr->fa_meta);
+	free(test_attr->fa);
+	free(test_attr);
+}
+
+/**
+ * Test if the init function works
+ */
+UK_TESTCASE(ukfallocbuddy, test_init)
+{
+	struct uk_falloc *fa;
+
+	UK_ASSERT(fa = malloc(uk_fallocbuddy_size()));
+	UK_TEST_EXPECT_SNUM_EQ(uk_fallocbuddy_init(fa), 0);
+
+	free(fa);
+}
+
+/**
+ * Test if the addmem functionality works
+ */
+UK_TESTCASE(ukfallocbuddy, test_add_memory)
+{
+	__sz len = 10000;
+	void *start_zone = malloc(len);
+	__vaddr_t dm_off;
+	__paddr_t start = (__paddr_t)start_zone;
+	unsigned long pages = len >> PAGE_SHIFT;
+	__sz fa_meta_size = uk_fallocbuddy_metadata_size(pages);
+	void *fa_meta = malloc(fa_meta_size);
+	struct uk_falloc *fa = malloc(uk_fallocbuddy_size());
+	int addmem_result;
+
+	uk_fallocbuddy_init(fa);
+
+	start = PAGE_ALIGN_UP(start + fa_meta_size);
+	pages = (len - fa_meta_size) >> PAGE_SHIFT;
+	dm_off = start;
+	addmem_result = fa->addmem(fa, fa_meta, start, pages, dm_off);
+	UK_TEST_EXPECT_SNUM_EQ(addmem_result, 0);
+
+	free(start_zone);
+	free(fa_meta);
+	free(fa);
```

Comment:
1. You mostly reuse here code that you have split in `initialize_parameters()` and `init_new_fb_alloc()`. I suggest to write it here in the same style as before so you recognize it is the same code. This makes is easier to maintain.
2. I increased the memory size that you add. You were adding less than 2.5 pages. Although it would be valid to have a test that explores the lower and upper limits.
3. You could argue that - as a best practice - you should release memory in the order that there is no internal references (in this case the exact reverse). But I kept it as-is since it is all released anyways. But you keep this in mind in other cases.

```suggestion
	__sz len;
	unsigned long pages;
	void *start_zone;
	__sz fa_meta_size;
	void *fa_meta;
	struct uk_falloc *fa;
	__paddr_t start;
	int rc;

	len = 10 * 1024 * 1024; /* 10 MiB */
	pages = len >> PAGE_SHIFT;
	start_zone = malloc(len);
	fab_bug_on(start_zone == NULL);

	fa_meta_size = uk_fallocbuddy_metadata_size(pages);
	fa_meta = malloc(fa_meta_size);
	fab_bug_on(fa_meta == NULL);

	fa = malloc(uk_fallocbuddy_size());
	fab_bug_on(fa == NULL);

	rc = uk_fallocbuddy_init(fa);
	fab_bug_on(rc != 0);

	start = PAGE_ALIGN_UP((__paddr_t)start_zone + fa_meta_size);
	pages = (len - fa_meta_size) >> PAGE_SHIFT;

	rc = fa->addmem(fa, fa_meta, start, pages, start);
	UK_TEST_EXPECT_ZERO(rc);

	free(start_zone);
	free(fa_meta);
	free(fa);
```